flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/sources/TableSourceUtil.scala

```diff
@@ -374,15 +374,17 @@ object TableSourceUtil {
         // push an empty values node with the physical schema on the relbuilder
         relBuilder.push(createSchemaRelNode(resolvedFields))
         // get extraction expression
-        resolvedFields.map(f => ResolvedFieldReference(f._1, f._3))
+        resolvedFields.map(f => new FieldReferenceExpression(f._1, f._3))
       } else {
-        new Array[ResolvedFieldReference](0)
+        new Array[FieldReferenceExpression](0)
       }
 
       val expression = tsExtractor.getExpression(fieldAccesses)
       // add cast to requested type and convert expression to RexNode
-      // TODO we cast to planner expressions as a temporary solution to keep the old interfaces
-      val rexExpression = Cast(expression.asInstanceOf[PlannerExpression], resultType)
+      val rexExpression = ApiExpressionUtils
```

Comment:
I would leave a TODO here as it reminds us that this is still not the final solution.